drivers/i2c_bitbang/i2c_master.c

```diff
@@ -0,0 +1,506 @@
+/* Copyright 2021 Andrei Purdea (portions)
+ * Copyright (C) 2019 Elia Ritterbusch (portions)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* This library follows the convention of the AVR i2c_master library.
+ * As a result addresses are expected to be already shifted (addr << 1).
+ */
+
+#include "i2c_master.h"
+#include "timer.h"
+#include "gpio.h"
+#include "wait.h"
+#include <stdbool.h>
+
+#ifndef I2C_BITBANG_SDA_PIN
+#    error "You must set I2C_BITBANG_SDA_PIN in config.h, in order to user the i2c bitbang driver"
+/* Define it to something, to avoid follow-up error messages from the compiler */
+#    define I2C_BITBANG_SDA_PIN 0
+#endif
+
+#ifndef I2C_BITBANG_SCL_PIN
+#    error "You must set I2C_BITBANG_SCL_PIN in config.h, in order to user the i2c bitbang driver"
+/* Define it to something, to avoid follow-up error messages from the compiler */
+#    define I2C_BITBANG_SCL_PIN 0
+#endif
+
+#ifndef I2C_BITBANG_ENABLE_INTERNAL_PULLUPS
+#    ifdef PAL_MODE_OUTPUT_OPENDRAIN
+#        define INITIALISE_PIN(pin) palSetLineMode((pin), PAL_MODE_OUTPUT_OPENDRAIN)
+#        define WRITE_PIN(pin, value) writePin((pin), (value))
+#        define READ_PIN(pin) readPin(pin)
+#    else
+#        define INITIALISE_PIN(pin) setPinInput(pin)
+#        define WRITE_PIN(pin, value) do { if (value) { setPinInput(pin); } else { writePin((pin), 0); setPinOutput(pin); } } while (0)
```

Comment:
If you really care about speed on AVR, it could be possible to provide an optimized implementation for AVR that manipulates just the `DDRx` register in `WRITE_PIN()` (assuming that the previous `INITIALIZE_PIN()` had set the `PORTx` register bit to 0). This would also help to avoid a potential conflict with software PWM emulation, similar to #5953 (although it might not be a problem for this particular code, because pins here are compile time constants, therefore the `DDRx` and `PORTx` bit manipulations may be compiled to single `cbi` or `sbi` instructions, which are atomic anyway).

Reply:
Alright, so I have tried this, but it had an opposite effect.
Instead, what I went ahead and did, was I implemented optimized versions of the routines for AVR, using inline assembly, that can achieve exact frequencies, as long as no interrupt interrupts us, and there is no clock stretching.